### PR TITLE
feat(mito2): add series index planning and builders

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Part 1 of 2. Base: `main`; head: `feat/series-index-builders`. Follow-up: #9086 adds background reconciliation and targets this branch. Merge this PR first, then rebase and retarget #9086 to `main`.

## What's changed and what's your intention?

Add bucket planning and SST-to-index builders as the foundation for background series-index maintenance. Planning merges overlapping time coverage, selects complete sequence suffixes, and accounts for TTL expiration. Builders produce companion range indexes and aggregate series indexes while reusing existing range coverage. Catalog loading reconstructs in-memory bucket coverage from existing entries.

Background maintenance remains inactive at this step. Existing public writer/searcher interfaces and catalog serialization are unchanged. Direct builder tests cover construction, reuse of a separately built range index, and preservation of completed range files when series construction fails. Catalog tests verify restored bucket coverage.

Validation after rebasing onto upstream `main` (`f3b29ee700`): `cargo nextest run -p mito2`: **1,348 passed** (one compaction test passed on retry). `make fmt` and `git diff --check` passed. Updated both series-index test request initializers with `skip_wal: false` to fix compilation against the current region request API.

Outstanding validation: full workspace tests, Clippy, unused-dependency checks, and Hawkeye. Local checks were limited by disk space and unavailable Docker/tool downloads. New Rust files retain the existing 2023 license header.

## PR Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.
- [ ] This PR needs to be backported to release branches, and I have added the `backport-<target>` labels.

### CODE REVIEW

File: src/mito2/src/series_index/bucket.rs [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+        {
+            continue;
+        }
+        // Select all SSTs above the watermark, including every file sharing a sequence.
+        // File IDs may change under compaction without advancing indexed coverage.
+        // The maximum remains valid because it exceeds the watermark and is retained.
+        bucket.files.retain(|file| {
+            file.meta_ref()
+                .sequence
+                .is_some_and(|sequence| sequence.get() > index_bucket.max_file_sequence)
```
Comment (WenyXu) [inline src/mito2/src/series_index/bucket.rs]: The merged maximum watermark does not establish coverage across the whole merged interval. In the bridge test, the sequence-15 SST spans `[10,30]`, but `[0,20)` is only indexed through sequence 10 and `[20,30)` has no existing coverage. Taking another bucket’s watermark of 30 silently excludes this SST.

Could we check coverage against the original intervals and their watermarks before merging, and update the bridge test to retain this uncovered source?

File: src/mito2/src/series_index/builder.rs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    let mut writer = SeriesIndexWriter::try_new(
+        region.metadata(),
+        store.clone(),
+        &path,
+        SeriesIndexWriterOptions::default(),
+        Some(series_metadata(entry)?),
+    )
+    .await?;
+    while let Some(batch) = visible.try_next().await? {
+        writer.write(&batch).await?;
```
Comment (WenyXu) [inline src/mito2/src/series_index/builder.rs]: These early returns drop unfinished writers without calling `abort()`, which both index writers explicitly require after a write error. Read failures can take the same path, and a combined build can also drop unfinished companion range writers, leaving temporary files or uploads behind.

Could we abort all unfinished writers on failure while preserving completed range indexes? The current close-failure test exercises `finish()`’s own cleanup; a mid-read/write failure test would cover this path.

File: src/mito2/src/series_index/builder.rs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        crate::sst::parquet::row_selection::RowGroupSelection,
+    )>,
+> {
+    Ok(region
+        .access_layer
+        .read_sst(file)
+        .projection(Some(ReadColumns::new([])))
+        .build_reader_input(&mut ReaderMetrics::default())
+        .await?
+        .map(|(context, selection)| (Arc::new(context), selection)))
```
Comment (WenyXu) [inline src/mito2/src/series_index/builder.rs:62]: This reader retains the SST’s schema without installing `FlatCompatBatch`, while the writers use region/version metadata. After widening the time index from milliseconds to microseconds, an old millisecond SST will therefore fail the writer’s timestamp-unit check.

Could we reuse the existing compatibility conversion before merging/writing, targeting `version.metadata` consistently, including for the series writer? A test mixing SSTs written before and after widening should verify the resulting min/max timestamps.

File: src/mito2/src/series_index/bucket.rs [hunk 4]
```diff
@@ -0,0 +1,772 @@
...
+            let end = file.time_range().1.split().0;
+            let sequence = file
+                .meta_ref()
+                .sequence
+                .map_or(0, |sequence| sequence.get());
+            let first_window = start.div_euclid(compaction_window_secs);
+            let last_window = end.div_euclid(compaction_window_secs);
+            let window_count = i128::from(last_window) - i128::from(first_window) + 1;
+            let window_sequences = if window_count > MAX_FILE_WINDOW_SEQUENCES as i128 {
+                BTreeMap::new()
```
Comment (WenyXu) [inline src/mito2/src/series_index/bucket.rs]: An SST spanning more than 32 windows produces an empty coverage map, but `to_series_entry()` still allows the build. Since the reuse check explicitly rejects empty maps, an unchanged bucket with at least four SSTs will produce another full replacement on every planning pass.

Could we defer these builds or retain a conservative way to recognize unchanged inputs? A repeated-planning test at the 32/33-window boundary would help cover this case.

File: src/mito2/src/series_index/bucket.rs [hunk 5]
```diff
@@ -0,0 +1,772 @@
...
+/// map is publishable only after every planned build and the catalog writes succeed.
+pub(crate) fn plan_series_indexes(
+    buckets: Vec<SeriesBucket>,
+    mut index_buckets: BTreeMap<Timestamp, IndexBucket>,
+    ttl: Option<TimeToLive>,
+    now_ms: i64,
+) -> SeriesIndexPlan {
+    // Reconciliation changes geometry, not established coverage. In particular,
+    // a deferred bridge must not change the indexed snapshot.
+    let buckets = reconcile_series_buckets(buckets, &mut index_buckets.clone());
```
Comment (WenyXu) [inline src/mito2/src/series_index/bucket.rs]: This clones every index ID and per-window coverage map, then merges those maps, although reconciliation only uses the resulting interval boundaries. The copied coverage is discarded afterward, and merged coverage maps have no size limit.

Could we construct the temporary map from `IndexBucket::new(bucket.start, bucket.end)` instead? That should preserve the established snapshot and reuse the existing reconciliation logic without copying unrelated coverage data.

File: src/mito2/src/series_index/catalog.rs [hunk 6]
```diff
@@ -41,9 +45,21 @@ pub(crate) struct SeriesIndexEntry {
...
+    /// Width used to align the half-open compaction windows, in seconds.
+    pub(crate) compaction_window_secs: i64,
+    /// Maximum source SST sequence per compaction window, persisted for recovery.
+    /// Keys are window starts in epoch seconds; each window spans
+    /// `[start, start + compaction_window_secs)`. An SST contributes to every window
+    /// its time range intersects. Windows without source SSTs are omitted.
+    /// An empty map means a source SST exceeded the per-file tracking limit and
+    /// cannot establish whether source data is already indexed. Merged coverage
+    /// has no entry limit.
+    pub(crate) window_sequences: BTreeMap<i64, u64>,
```
Comment (WenyXu) [inline src/mito2/src/series_index/catalog.rs]: These new fields are required during deserialization, so catalogs written with the previous layout will fail to load and fall back to an empty series catalog. Background construction is not active yet, but this still changes the persisted format without an old-layout recovery test.

Could we add `#[serde(default)]` to both fields and test loading the previous layout? The existing zero-width/empty-map semantics already provide a conservative fallback that does not establish reuse.

Comment (evenyag) [reply src/mito2/src/series_index/catalog.rs]: Breaking is allowed.

File: src/mito2/src/series_index/builder.rs [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+            reason: "series-index source disappeared",
+        })?
+    } else {
+        Box::pin(
+            FlatMergeReader::new(schema, sources, 8192, None)
+                .await?
+                .into_stream(),
+        )
+    };
+    // TODO: Deduplicate update-mode rows before series indexes are used by queries.
```
Comment (WenyXu) [inline src/mito2/src/series_index/builder.rs]: nit:
```suggestion
    // TODO(yingwen): Deduplicate update-mode rows before series indexes are used by queries.
```

**Review by WenyXu — APPROVED** [review]
REST LGTM
